src/confd/src/ietf-system.c

```diff
@@ -1018,6 +1003,67 @@ static int change_auth(sr_session_ctx_t *session, uint32_t sub_id, const char *m
 	return SR_ERR_OK;
 }
 
+static int change_nacm(sr_session_ctx_t *session, uint32_t sub_id, const char *module,
+		       const char *xpath, sr_event_t event, unsigned request_id, void *priv)
+{
+	sr_val_t *users = NULL, *groups = NULL, *rules = NULL;
+	size_t user_count = 0, group_count = 0, rule_count = 0;
+	struct confd *confd = (struct confd *)priv;
+	int rc;
+
+	if (event != SR_EV_DONE)
+		return SR_ERR_OK;
+
+	/* Fetch all users from ietf-system */
+	rc = sr_get_items(session, XPATH_AUTH_"/user/name", 0, 0, &users, &user_count);
+	if (SR_ERR_OK != rc) {
+		ERROR("Failed fetching system users: %s", sr_strerror(rc));
+		goto cleanup;
+	}
+
+	aug_load(confd->aug);
+
+	for (size_t i = 0; i < user_count; i++) {
+		const char *user = users[i].data.string_val;
+		char xpath_groups[256];
+		bool is_admin = false;
+
+		/* Fetch groups for each user */
+		snprintf(xpath_groups, sizeof(xpath_groups), NACM_BASE_"/groups/group[user-name='%s']/name", user);
+		rc = sr_get_items(session, xpath_groups, 0, 0, &groups, &group_count);
+		if (!rc) {
+			for (size_t j = 0; j < group_count; j++) {
+				char xpath_rules[256];
+
+				/* Fetch and check rules for each group */
+				snprintf(xpath_rules, sizeof(xpath_rules), NACM_BASE_"/rule-list"
+					 "[group='%s']/rule[module-name='*'][access-operations='*']"
+					 "[action='permit']", groups[j].data.string_val);
+				rc = sr_get_items(session, xpath_rules, 0, 0, &rules, &rule_count);
+				if (rc)
+					continue;
+
+				/* At least one group grants full administrator permissions */
+				is_admin = true;
+				sr_free_values(rules, rule_count);
+			}
+			sr_free_values(groups, group_count);
+		}
+
+		if (is_admin)
+			systemf("adduser %s wheel", user);
+		else
+			systemf("delgroup %s wheel", user);
```

Comment:
What is the mechanism that ensures that `user` exists here?

I.e., when loading a startup with user `foo`, how are we sure that `sys_add_user()` has already run? Conversely, when removing `foo` and its associated ACL, how are we sure that `sys_del_user()` has _not_ already run?

Reply:
Yeah, there really isn't any, except for this monitor running at lower prio than the primary one used by sysrepo/netopeer2.  This is of course no guarantee to the ordering between it and ietf-system.yang. 

I considered adding a must expression to at least ensure users exist in ietf-system before adding them?  We can discuss it over a cup o' joe in the office in a bit :coffee: :croissant:

Reply:
Decisions from AFK coffee meeting:

 1. Check BusyBox impl. of `adduser` and `delgroup`, are they just manipulating strings in the `/etc/group` file, or do they verify the user exists in `/etc/passwd` as well?
 2. Check return value of `systemf()` and log error

Reply:
1. OK, so `adduser` and `delgroup` *do* check if the user exists in `/etc/passwd`.  So we need to scrap them and try again with augeas
2. Check return value of augeas

Reply:
Fixed with augeas. Dunno why I couldn't get it to work previously, but now it works at least. And great, I might add!